### PR TITLE
refactor(meta): store compaction group info in HummockVersion

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

#6246 has made update of compaction group info(e.g. member tables) and hummock version info in sync by using one transaction. But due to the fact that compaction group info and hummock version info are managed in different places (`CompactionGroupManagerInner` and `HummockManager`), the code to update them has been hard to read/maintain to some extent. For example, 

- `CompactionGroupManagerInner` is a member of `HummockManager`, but its methods [take ref of `HummockManager`](https://github.com/risingwavelabs/risingwave/blob/037f51ae7c9608e68d8697f7f87b0e188a78f161/src/meta/src/hummock/manager/compaction_group_manager.rs#L350).
- [`sync_group method`](https://github.com/risingwavelabs/risingwave/blob/037f51ae7c9608e68d8697f7f87b0e188a78f161/src/meta/src/hummock/manager/mod.rs#L1469)'s parameters `compaction`,`versioning`,`compaction_groups` actually implies 3 corresponding locks have been taken. An unconscious locking request inside the method will cause dead lock.
- [`sync_group`] takes a `trx_extern_part`. The external caller of `sync_group` must remember to check `trx_extern_part` after `sync_group` returns, in order to make sure the external trx is committed, if it has not been done by `sync_group`.
- The lock order of `CompactionGroupManagerInner` and `versioning` need be maintained carefully.

This PR wants to simplify the code, by 
- store compaction group info(except compaction config) in hummock version directly. We remove the universal [`sync_group`], and split its functionality into `register_table_ids`/`unregister_table_ids`. Note that because we haven't supported compaction group split API, some functionality of [`sync_group`] may not be kept. For example [`sync_group`] handles SST split but current codebase never trigger it when calling [`sync_group`] .
- make `CompactionGroupManager` only responsible for compaction config maintenance now, which doesn't need to be in sync with hummock version.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

**Comment by zwang28** [comment]
@hzxa21 @soundOfDestiny @Li0k 
The majority of this refactor has been done. PTAL if possible.

**Review by soundOfDestiny — APPROVED** [review]
current part is trivial

File: src/meta/src/rpc/service/hummock_service.rs [hunk 1]
```diff
@@ -444,17 +427,18 @@ where
...
-            .await
-            .iter()
-            .map(|cg| cg.into())
-            .collect_vec();
-        Ok(Response::new(RiseCtlListCompactionGroupResponse {
-            status: None,
-            compaction_groups,
-        }))
+        todo!("#7817")
+        // let compaction_groups = self
```
Comment (Li0k) [inline src/meta/src/rpc/service/hummock_service.rs:431]: remove this code ?

Comment (zwang28) [reply src/meta/src/rpc/service/hummock_service.rs]: fixed

File: src/meta/src/hummock/manager/compaction_group_manager.rs [hunk 2]
```diff
@@ -18,92 +18,72 @@ use std::ops::DerefMut;
...
 
-    pub async fn compaction_group(&self, id: CompactionGroupId) -> Option<CompactionGroup> {
-        self.compaction_group_manager
-            .read()
-            .await
-            .compaction_groups
-            .get(&id)
-            .cloned()
+    #[named]
+    pub async fn compaction_group_ids(&self) -> Vec<CompactionGroupId> {
```
[the remark this answers is not in the document]
Comment (zwang28) [reply src/meta/src/hummock/manager/compaction_group_manager.rs:85]: Make sense.
- Added comments.
- Removed its usage inside hummock manager.

**Comment by zwang28** [comment]
@StrikeW FYI after this PR, table register/unregister are also kept in version delta, so compaction test doesn't need to explicitly register compaction groups at the beginning of the test.

File: src/meta/src/hummock/manager/compaction_group_manager.rs [hunk 3]
```diff
@@ -18,92 +18,72 @@ use std::ops::DerefMut;
...
-            .compaction_groups
-            .values()
-            .cloned()
-            .collect_vec()
-    }
-
-    pub async fn compaction_groups_and_index(
+    /// Gets compaction config for `compaction_group_id` if exists, or returns default.
+    #[named]
+    pub async fn get_compaction_config(
```
[the remark this answers is not in the document]
Comment (zwang28) [reply src/meta/src/hummock/manager/compaction_group_manager.rs:73]: Currently all callers of `get_compaction_config` are aware of `compaction_group_id`. May apply this suggestion when there is an use case.

**Comment by zwang28** [comment]
Trying to remove `table_id_to_options` from version/version_delta.

**Review by Li0k — APPROVED** [review]
LGTM, thanks for the great work

**Pull request merged**